server/plex_sync.go

```diff
@@ -0,0 +1,149 @@
+package main
+
+import (
+	"errors"
+	"log/slog"
+	"time"
+
+	"gorm.io/gorm"
+)
+
+type PlexSyncResponse struct {
+	JobId string `json:"jobId"`
+}
+
+// Perform a Plex sync.
+// Errors are added silently to the job.
+func plexSyncWatched(
+	db *gorm.DB,
+	jobId string,
+	userId uint,
+	userThirdPartyAuth string,
+) {
+	updateJobCurrentTask(jobId, userId, "fetching libraries")
+	libraries, err := plexGetLibraries(userThirdPartyAuth)
+	if err != nil {
+		addJobError(jobId, userId, "failed to get plex libraries")
+		updateJobStatus(jobId, userId, JOB_DONE)
+		return
+	}
+	for _, library := range libraries {
+		if library.Type == string(MOVIES) {
+			updateJobCurrentTask(jobId, userId, "importing movies")
+			movies, err := plexGetLibraryItems(userThirdPartyAuth, library.Key)
+			if err != nil {
+				slog.Error("Failed to fetch movies from library", "library", library.Key, "error", err)
+				addJobError(jobId, userId, "failed to fetch movies from library "+library.Key)
+				continue
+			}
+			for _, movie := range movies {
+				if movie.ViewCount == 0 && movie.UserRating == 0 {
+					// Not viewed and not rated, skip importing
+					continue
+				}
+				updateJobCurrentTask(jobId, userId, "importing movie "+movie.Title)
+				slog.Info("plexSyncWatched: Importing movie.", "movie_name", movie.Title, "user_id", userId)
+				sr, err := searchMovie(movie.Title, string(movie.Year))
+				if err != nil {
+					slog.Error("Failed to search for movie", "movie", movie.Title, "error", err)
+					addJobError(jobId, userId, "failed to search for movie "+movie.Title)
+					continue
+				}
+
+				if len(sr.Results) == 0 {
+					slog.Error("plexSyncWatched: could not find movie", "movie", movie.Title)
+					addJobError(jobId, userId, "failed to add movie "+movie.Title)
+					continue
+				}
+				// We have to assume it's the first result. Thank you Plex, for not giving better metadata
+				// In theory, we could grab the IMDB ID and cross reference it with TMDB's /movie/external_ids API,
+				// but that seems like a lot of overhead. Generally title + year is good enough.
+				match := sr.Results[0]
+				_, err = addWatched(db, userId, WatchedAddRequest{
+					Status:      FINISHED,
+					ContentID:   match.ID,
+					ContentType: MOVIE,
+					Rating:      int8(movie.UserRating),
+					WatchedDate: time.Unix(int64(movie.LastViewedAt), 0),
+				}, IMPORTED_WATCHED_PLEX)
+				if err != nil {
+					if err.Error() == "content already on watched list" {
```

Comment:
I think here we should attempt to update the rating, thoughts @IRHM ?

Reply:
Good idea @stephaje, but I think that might cause problems where a user changes their rating on Watcharr, then a full re-sync happens with plex and their rating is overridden.

I suppose we could do a two-way sync with plex to get around that, but that sounds pretty complex. I would imagine most people only sync, then only interact with the show (for ratings/status) through watcharr (not sure)?

Reply:
I would imagine most people would resync every couple of weeks to pull in fresh watches. If they watched something but hadn't rated it when they did the first sync, then rated it in plex, then resync, it would be nice to pull in. Maybe it will update only if the current watcharr rating is empty?

Reply:
Sounds good @stephaje. As long as it can't accidentally override a rating set in watcharr it should be alright.